### PR TITLE
Fixing test and benign race condition during table drop

### PR DESCRIPTION
**Pull request by alesapin**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix race condition which may lead to an exception during table drop. It's a bit tricky and not dangerous at all. If you want an explanation, just notice me in telegram.

### CODE REVIEW

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -2997,6 +2996,12 @@ void StorageReplicatedMergeTree::shutdown()
         std::unique_lock lock(data_parts_exchange_endpoint->rwlock);
     }
     data_parts_exchange_endpoint.reset();
+
+    /// We clear all parts after stopping all background operations. It's
+    /// important, because background operations can produce temporary parts
+    /// which will remove themselfs in their descrutors. If so, we may have race
+    /// condition between our remove call and background process.
+    clearOldPartsFromFilesystem(true);
```
Comment (tavplubix) [inline src/Storages/StorageReplicatedMergeTree.cpp:3004]: Why not to do the same for [StorageMergeTree](https://github.com/ClickHouse/ClickHouse/blob/c7d9094a7a42488a84b8ed463bd9d2655a9f1c02/src/Storages/StorageMergeTree.cpp#L126)?

File: tests/queries/0_stateless/01079_parallel_alter_modify_zookeeper.sh [hunk 2]
```diff
@@ -100,8 +100,14 @@ wait
 
 echo "Finishing alters"
 
-# This alter will finish all previous, but replica 1 maybe still not up-to-date
-while [[ $(timeout 30 $CLICKHOUSE_CLIENT --query "ALTER TABLE concurrent_alter_mt_1 MODIFY COLUMN value1 String SETTINGS replication_alter_partitions_sync=2" 2>&1) ]]; do
+# This alter will finish all previous, but replica 1 maybe still not up-to-date.
+# If query will throw something, than we will sleep 1 and retry. If timeout
+# happened we will silentrly go out of loop and probably fail tests in the
```
Comment (tavplubix) [inline tests/queries/0_stateless/01079_parallel_alter_modify_zookeeper.sh:105]: Typo: silently

**Review by tavplubix — APPROVED** [review]

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 3]
```diff
@@ -2997,6 +2996,12 @@ void StorageReplicatedMergeTree::shutdown()
         std::unique_lock lock(data_parts_exchange_endpoint->rwlock);
     }
     data_parts_exchange_endpoint.reset();
+
+    /// We clear all parts after stopping all background operations. It's
+    /// important, because background operations can produce temporary parts
+    /// which will remove themselfs in their descrutors. If so, we may have race
```
Comment (alexey-milovidov) [inline src/Storages/StorageReplicatedMergeTree.cpp:3002]: themselves

File: tests/queries/0_stateless/01079_parallel_alter_modify_zookeeper.sh (hunk 2 above)
Comment (alexey-milovidov) [inline tests/queries/0_stateless/01079_parallel_alter_modify_zookeeper.sh:105]: silentrly -> silently

**Pull request merged**